service/history/workflowResetor.go

```diff
@@ -881,3 +887,141 @@ func (w *workflowResetorImpl) replicateResetEvent(baseMutableState mutableState,
 	newMsBuilder.UpdateReplicationStateLastEventID(clusterMetadata.GetCurrentClusterName(), lastEvent.GetVersion(), lastEvent.GetEventId())
 	return
 }
+
+func (w *workflowResetorImpl) AutoResetWorkflow(currWorkflowContext workflowExecutionContext, currMutableState mutableState) (retError error) {
```

Comment:
Is this logic only called through transfer task?  Can we separate out decision to reset from actual reset into two separate piece of logic.  I expect decision to reset to be really fast which does not need any database reads.

Reply:
Reset need to read database (current/base mutableState + base history) and have to be expensive. Either reading them in transferTask or inside this reset logic.
The only way to save perf is to reduce the chance of reset  failures because of:
1) base deleted
2) base not resettable
3) current not resettable
4) domainConfig change